Add Timezone Configuration Option for Logging

Nabstar3 opened the pull request:
## Description

**Closes #130** 

This pull request introduces a new feature that allows users to specify the timezone for logging and debugging purposes. This enhancement aims to provide greater flexibility and accuracy in log timestamps, catering to users who wish to use a custom timezone, facilitating easier log analysis and debugging.

## Changes Made

- Added a new configuration option in the settings to specify the timezone.
- Modified the logging setup to use the specified timezone for timestamps.
- Updated the documentation to include instructions on how to set the timezone option.

## Benefits

- **Better support for Docker**: Allows users running a base image in a docker container to now have the option to include timestamps which were not previously available.
- **Enhanced Flexibility**: Supports a wide range of timezones, accommodating users from various geographical locations.
- **Better Debugging**: Simplifies the debugging process by aligning log timestamps with local time, especially useful in distributed systems.

## How to Use

To specify the timezone for logging, update the environment variable within the `docker-compose.yaml` with the desired timezone string. For example:

```json
  TIME_ZONE: "America/New_York"
```

Ensure that the timezone string is valid and recognized by the system. This feature leverages the IANA Time Zone Database for timezone identification.

## Impact

This change is backward compatible and does not affect existing functionality. Users who do not specify a timezone will default to the system's timezone.

---

ManiMatter commented on docker-compose.yaml:
```diff
@@ -0,0 +1,48 @@
+    container_name: decluttarr
+    restart: always
+    environment:
+      - TIME_ZONE=Europe/Zurich
```
I would stick with "TZ"  since that is - to my knowledge - a docker standard. However, I am not sure whether setting TZ actually chagned the timezone of the decluttarr image.

On a quick google search, many articles talk about `tzdata` package that needs to be installed for ubuntu-based images (which decluttarr is, as you will see in the workflow-yml files). 

Here's smth I found: https://dev.to/bitecode/set-timezone-in-your-docker-image-d22#comment-1h7e2

Do you want to have a look? I think that's relevant since when printing out the timestamps I guess we want the time to be correct ;)

Nabstar3 replied:
I did have a look at the link you mentioned.

The overall consensus was to first install the `tzdata` package and then add the following to your docker file:

```yaml
FROM ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive

ENV TZ Europe/Berlin

RUN apt-get update \
&& apt-get install -yq tzdata locales \
&& rm -rf /var/lib/apt/lists/* \

&& sed -i '/en_US.UTF-8/s/^# //g' /etc/locale.gen \
&& locale-gen \
&& ln -fs /usr/share/zoneinfo/${TZ} /etc/localtime \
&& dpkg-reconfigure tzdata

ENV LANG en_US.UTF-8

ENV LANGUAGE en_US:en
ENV LC_ALL en_US.UTF-8

CMD ["/bin/bash"]
```

At the minimum, this might work:

```
FROM alpine:latest

RUN apk update && apk add tzdata
ENV TZ="America/New_York"
```

It is interesting to note this is not an issue on debian based images.

ManiMatter replied:
I don't really see a reason why we couldn't change to from ubuntu to alpine. 

> At the minimum, this might work:
> 
> ```
> FROM alpine:latest
> 
> RUN apk update && apk add tzdata
> ENV TZ="America/New_York"
> ```

If this gives us correct system time stettings, I think we should go for it

Nabstar3 replied:
> I don't really see a reason why we couldn't change to from ubuntu to alpine

I don't know enough about docker to comment on the difference between the two. But I did find out that `python:3.10.13-slim` is actually not ubuntu but debian...

The change is not neccessary as my tests revealed that `tzdata` is already a package installed in `python:3.10.13-slim`. All I did was upgrade it. The only neccessary changes are what I mentioned in my comment `Proposed Solution` below

ManiMatter commented on docker-compose.yaml:
```diff
@@ -0,0 +1,48 @@
+      - LOG_LEVEL=INFO
+      #- TEST_RUN=True
+      #- SSL_VERIFICATION=False
+      ## Features
```
I would not use the TZ variable to steer whether or not the timestamps are included in the logs. I think we should add a explicit setting for that, e.g. "`LOGS_WITH_TIMESTAMP`", defaulting to false and being an optional parameter.

Nabstar3 replied:
See my comment below

ManiMatter commented on README.md:
```diff
@@ -55,19 +63,19 @@ services:
     container_name: decluttarr
     restart: always
     environment:
-      - TZ=Europe/Zurich
```
I would leave TZ, since that is a docker standard. 
See my comment further down below

ManiMatter commented on README.md:
```diff
@@ -3,6 +3,7 @@
 # **Decluttarr**
 
 ## Table of contents
+
```
This PR is difficult to review, since you also ran a format cleanup, which now marks many files & lines as changed, taking away attention from the relevant changes.

Could you please revert that, so we only see the material changes?

ManiMatter commented on README.md:
```diff
@@ -123,31 +135,42 @@ Configures the general behavior of the application (across all features)
 - Permissible Values: True, False
 - Is Mandatory: No (Defaults to True)
 
+**TIME_ZONE**
```
As mentioned in the other comments - I suggest not to include that (and stick with the default TZ, which don't think explanation in my opinion), and rather to include a variable based on which we turn on or off the inclusion of timestamps in logs.

ManiMatter commented on src/utils/loadScripts.py:
```diff
@@ -1,180 +1,362 @@
 
+import logging
+from datetime import datetime
+import pytz  # Requires installing the pytz module
```
If we are correctly changing the system time of the container that runs this script (as suggested above), then I think we don't need to use pytz to change the timezone again within the python script, and we can simply rely on the system time. I think much of this code is the not needed

Nabstar3 replied:
While I agree that creating a custom `logging` class just to incorporate timestamps does involve more lines than one would think, the additions to the `Dockerfile` from the comment above involving `tzdata` is also quite _ugly_ and also faces the same length issue as the implementation here.

ManiMatter replied:
If we changed to alpine, it take it the addition would be very small on the .yml files. Would you agree we then don't need a custom class?

Nabstar3 replied:
I agree we dont need a custom class, however see my comment above about changing to alpine

ManiMatter commented on src/utils/loadScripts.py:
```diff
@@ -1,180 +1,362 @@
+
 def setLoggingFormat(settingsDict):
-    # Sets logger output to specific format
-    log_level_num=logging.getLevelName(settingsDict['LOG_LEVEL'])
```
If we included the specific setting that defines whether to print out or not the timestamp, I think all we had to do is to add a if/else statement here to include it, wouldn't we?

Nabstar3 replied:
I think that timestamps are an expected compliment to logs globally. 

I can not foresee anyone complaining of timestamps being incorporated into system logs and I'm not sure if it's worth introducing a toggle feature?

If this is dealbreaker however, I'm happy to accomodate this request.

ManiMatter replied:
> I can not foresee anyone complaining of timestamps being incorporated into system logs and I'm not sure if it's worth introducing a toggle feature?

when you look at the logs through dozzle, the timestamp is already there. Thus if we added it as you suggest (without toggle), wouldn‘t the timestamp show twice?

ManiMatter commented on src/utils/loadScripts.py:
```diff
@@ -1,180 +1,362 @@
-
-    logger.info('') 
-    return   
+    logger.info(
```
As written below - pls revert the format changes. I know it doesn't uphold currently to certain coding standards (ie too long lines etc), but I feel it was easier to read before.

Nabstar3 replied:
Will try. Hopefully it's easy to unformat.

ManiMatter replied:
What I would do is:
- Copy your branch
- Revert your dev branch with git rebase -i XXX
- Delete the commits where the cleanup took place
- git push force
- then merge your copied branch into dev branch (either manually by copying over the relevant code, or litterally a merge where you don't merge the format changes)


maybe there's a more clever way..

Nabstar3 commented:
## Summary

After reviewing the [link](https://dev.to/bitecode/set-timezone-in-your-docker-image-d22#comment-1h7e2), I performed the following steps:

1. Added `TZ=EST` to the `docker-compose.yaml`.
2. Rebuilt the container.
3. Upgraded the `tzdata` package inside the container.
4. Modified `loadScripts.py` to include `asctime` regardless of whether it's running in a Docker container, then restarted the container to verify accurate time zone reflection.
5. Restarted the container again.
6. Checked the logs.

The desired results were achieved, and timestamps are now accurately included in the logs:

`2024-07-21 02:57:40,340 [INFO]: LOG_LEVEL = INFO: Only logging changes (switch to VERBOSE for more info)`


**Note**: If a `TZ` value is not provided, the timestamp will display the correct date but the incorrect timezone, as explained [here](https://dev.to/bitecode/set-timezone-in-your-docker-image-d22#comment-1h7e2:~:text=The%20timezone%20is%20correct%2C%20but%20current%20time%20is%20incorrect).

## Thoughts

I am curious about the developer's decision to omit timestamps when running inside a Docker container. Two possible reasons come to mind:

1. They might have assumed that an unhandled error could occur if no system time was found in a Docker container.
2. They might have observed inaccurate timestamps when a valid `TZ` value was not provided and chose not to make the `TZ` environment variable mandatory, thus disabling it for Docker containers.

## Next Steps

Based on these observations, we can consider modifying the `setup_logging` function in `loadScripts.py` to not filter out timestamps based on the environment.

Considering my earlier thoughts:

> I believe that timestamps are generally expected in logs worldwide.
>
> I cannot foresee any complaints about incorporating timestamps into system logs, and I'm unsure if it's worth introducing a toggle feature for this.

Please let me know your thoughts on this matter, and we can discuss how to proceed.

ManiMatter commented:
> am curious about the developer's decision to omit timestamps when running inside a Docker container. Two possible reasons come to mind:
> 
> 1. They might have assumed that an unhandled error could occur if no system time was found in a Docker container.
> 2. They might have observed inaccurate timestamps when a valid `TZ` value was not provided and chose not to make the `TZ` environment variable mandatory, thus disabling it for Docker containers.

The reason I did not include the log timestamp in the log message is because the log timestamp is present anyways. 
a. With `docker-compose logs -t | grep 'decluttarr'`, you get the timestamp (although not neatly formatted)
b. When looking at the logs via a log viewer (portainer or dozzle, for instance), it is there:

Dozzle:
<img width="948" alt="image" src="https://github.com/user-attachments/assets/9e97c607-3e81-4d51-939d-f8e806d08fba">

Portainer (note the option to switch on display of timestamp at the top):
<img width="803" alt="image" src="https://github.com/user-attachments/assets/f7eb0cc0-1202-426e-ac2a-46ddca80d6ac">



If we included the timestamp in the message, it would most likely mean that the timestamp shows twice, which I would deem a bug. That's why I argue: If we are to include the timestamp in the message, then that should be an optional setting that is disabled by default.

Nabstar3 commented:
> If we included the timestamp in the message, it would most likely mean that the timestamp shows twice, which I would deem a bug. That's why I argue: If we are to include the timestamp in the message, then that should be an optional setting that is disabled by default.

While I can't validate the issue you're concerned about in terms of the double timestamps, I am willing to go with your proposed solution. 

However, I don't think think that dependency on a third party tool such as dozzle or portainer is expected nor assumed from the user, and it might be more well designed to have essential timestamps featured in logs as a standalone feature. I, for example, have never heard of dozzle before you introduced me to it and I think its fair to assume that future users may not have either. If there were a double timestamp, I don't believe it would be deemed as a bug as I'm sure `decluttarr` wasn't developed with the use of dozzle or other services in mind.

I think meeting in the middle with:

> If we are to include the timestamp in the message, then that should be an optional setting that is disabled by default.

Is valid, but as I mentioned above I think it's essential to make this a standalone feature and to not assume the use of dozzle or other third party tools, and thus we should **enable** this feature by default rather then **disable**.

##Proposed Solution

I think we're closing in on a solution. I'm willing to:

1. Rebase back to remote `dev`
2. Introduce a optional environment variable named `INCLUDE_TIMESTAMP` which defaults to `True` and is dependent on `TZ`.
3. Include your proposed `if/else` statement in `loadScripts.py` to check for the toggle condition and appropriately include `asctime`

Let me know your thoughts on the above.

ManiMatter commented:
> I think meeting in the middle with:
> 
> > If we are to include the timestamp in the message, then that should be an optional setting that is disabled by default.
> 
> Is valid, but as I mentioned above I think it's essential to make this a standalone feature and to not assume the use of dozzle or other third party tools, and thus we should **enable** this feature by default rather then **disable**.


I see it differently and would still argue it should be disabled by default.. 

My point is not about the 3rd party tools (ie portainer, dozzle). Rather, by default in the terminal the timestamp is **not** included. Only with the parameter `-t` that has to be explicitly set by the user it is shown. Thus making this feature disabled by default makes it the behavior consitent with the onboard terminal behavior. I am aware that certain apps bake the timestamp also into the message, but that then a) deprives the user of the possibility to look at the logs without timestamp (which is foreseen by option `-t`), and results in the double timestamp in downstream apps. 

Example:

<img width="909" alt="image" src="https://github.com/user-attachments/assets/1cb896a5-3eff-4317-96e8-648d32ecd16d">

Note the timestamp differences. The one by dozzle is correct, the one within the message is the systemtime of the container (which does not correctly adjust by setting TZ).

ManiMatter commented:
Hi,

I have thought about this a bit more and performed some tests.

I run this command to show the current time inside the container: `sudo docker exec -it decluttarr /bin/bash -c "date"`
When I run the container with `TZ: "America/New_York"` or `TZ: "Europe/Zurich"`, I get - as expected, two different (and correct) timestamps.

Therefore, I would argue that the above discussions we were having about correct/incorrect timestamp within the ubuntu-image is void. It's correct as it is.

Now to the other point: Should the timestamp be included in the logs?
Here, I would remain firm. 
When the script runs as a standalone python script, then the timestamp is already included.
When the script runs as docker image, then the logs can be retrieved in two ways:
a) Via a convenient 3rd party image that shows the logs (dozzle, portainer, etc), with the (correct) timestamp
b) Via `sudo docker logs -t decluttarr`, which gives the (correct) timestamp, although the format is not easily readable but it's a standard machine timestamp (Quote from ChatGPT: RFC3339 format, which is a profile of the ISO 8601 standard for date and time. This format is used widely in computing and data interchange). 

I would not carry on with implementing this; I think it doesn't add value. 
If you want convenience (e.g. nicely formatted timestamps), suggest you look into the tools I mentioned under a). I think you'll like dozzle a lot.

If you want to just have a quick look what has happened, the command b) should help, and even if the time stamp is not beautifully formatted, you still get the info what happened when.
